src/heuristic/HeuristicMapper.cpp

```diff
@@ -1025,198 +1076,568 @@ void HeuristicMapper::Node::applyTeleportation(const Edge&   swap,
+        // forwardMult > 0 && reverseMult == 0
+        swapCost = architecture->distance(physQ1, physQ2);
+      } else {
+        // forwardMult > 0 && reverseMult > 0
+        swapCost = std::max(architecture->distance(physQ1, physQ2),
+                            architecture->distance(physQ2, physQ1));
+      }
+      costHeur += swapCost;
     }
-    // 2-qubit-gates not yet mapped next to each other are handled in the
-    // heuristic
-  } else {
-    for (auto& swapNode : swaps) {
-      for (auto& swap : swapNode) {
-        if (swap.op == qc::SWAP) {
-          if (arch.bidirectional()) {
-            costFixed += COST_BIDIRECTIONAL_SWAP;
-          } else {
-            costFixed += COST_UNIDIRECTIONAL_SWAP;
-          }
-        } else if (swap.op == qc::Teleportation) {
-          costFixed += COST_TELEPORTATION;
-        }
+  }
+
+  return costHeur;
+}
+
+double HeuristicMapper::heuristicGateCountSumDistanceMinusSharedSwaps(
+    std::size_t layer, Node& node) {
+  if (node.validMapping) {
+    return 0.;
+  }
+  const auto& twoQubitGateMultiplicity = twoQubitMultiplicities.at(layer);
+  double      costHeur                 = 0.;
+  double      costReversals            = 0.;
+  std::vector<std::size_t> nSwaps{};
+  nSwaps.reserve(twoQubitGateMultiplicity.size());
+
+  for (const auto& [edge, multiplicity] : twoQubitGateMultiplicity) {
+    const auto& [q1, q2]                  = edge;
+    const auto [forwardMult, reverseMult] = multiplicity;
+    const auto physQ1 = static_cast<std::uint16_t>(node.locations.at(q1));
+    const auto physQ2 = static_cast<std::uint16_t>(node.locations.at(q2));
+
+    if (architecture->unidirectional()) {
+      // only for purely unidirectional architectures is it certain that at
+      // least one of the two directions has to be reversed
+      costReversals +=
+          std::min(forwardMult, reverseMult) * COST_DIRECTION_REVERSE;
+    }
+
+    if (node.validMappedTwoQubitGates.find(edge) !=
+        node.validMappedTwoQubitGates.end()) {
+      // validly mapped 2-qubit-gates
+      continue;
+    }
+
+    double swapCost = 0.;
+    if (forwardMult == 0) {
+      // forwardMult == 0 && reverseMult > 0
+      swapCost = architecture->distance(physQ2, physQ1, false);
+    } else if (reverseMult == 0) {
+      // forwardMult > 0 && reverseMult == 0
+      swapCost = architecture->distance(physQ1, physQ2, false);
+    } else {
+      // forwardMult > 0 && reverseMult > 0
+      swapCost = std::min(architecture->distance(physQ1, physQ2, false),
+                          architecture->distance(physQ2, physQ1, false));
+    }
+    costHeur += swapCost;
+
+    // infer maximum number of swaps in this distance
+    if (architecture->unidirectional()) {
+      nSwaps.emplace_back(
+          static_cast<std::size_t>(swapCost / COST_UNIDIRECTIONAL_SWAP));
+    } else {
+      nSwaps.emplace_back(
+          static_cast<std::size_t>(swapCost / COST_BIDIRECTIONAL_SWAP));
+    }
+  }
+
```

Comment:
Just a few lines below you use `architecture->bidirectional` for the if condition. This seems a little inconsistent and is probably one of the reasons why I originally brought it up in the comment for the changes in the `Architecture` class.

Reply:
Most of this probably cleared up with the discussion of semi-directionality. The use of both ` unidirectional()` and `bidirectional()` is intentional to account for semi-directional architectures.

This heuristic relies on subtracting the maximum cost savable up to the optimal solution via shared swaps.
So when calculating the number of swaps in each distance, we need to assume the cheapest possible swap cost for the highest possible number of swaps (which only in the case of unidirectional architectures is the higher `COST_UNIDIRECTIONAL_SWAP`)
Later we subtract the highest cost resulting from a given number of shared swaps, where we need to assume the _highest_ possible cost per swap (which only for bidirectional architectures is the lower `COST_BIDIRECTIONAL_SWAP`)

So in one case semi-directional architectures are handled together with bidirectional ones (since that is the worst case), and in another there are handled together with unidirectional architectures (since that is the worst case there)

Reply:
Yeah. That is all clear now. And now that o understand it better, it is quite elegant 😌

Reply:
Thanks😊